state/protocol/protocol_state/mutator_test.go

```diff
@@ -1,72 +1,289 @@
-	candidate := unittest.BlockHeaderFixture()
-	s.protocolStateDB.On("ByBlockID", candidate.ParentID).Return(nil, storerr.ErrNotFound)
-	updater, err := s.mutator.CreateUpdater(candidate)
-	require.ErrorIs(s.T(), err, storerr.ErrNotFound)
-	require.Nil(s.T(), updater)
+// TestHappyPathNoChanges tests that stateMutator doesn't cache any db updates when there are no changes.
+func (s *StateMutatorSuite) TestHappyPathNoChanges() {
+	s.params.On("EpochFallbackTriggered").Return(false, nil)
+	parentState := unittest.ProtocolStateFixture()
+	s.stateMachine.On("ParentState").Return(parentState)
+	s.stateMachine.On("Build").Return(parentState.ProtocolStateEntry, parentState.ID(), false)
+	err := s.mutator.ApplyServiceEvents([]*flow.Seal{})
+	require.NoError(s.T(), err)
+	hasChanges, updatedState, updatedStateID, dbUpdates := s.mutator.Build()
+	require.False(s.T(), hasChanges)
+	require.Equal(s.T(), parentState.ProtocolStateEntry, updatedState)
+	require.Equal(s.T(), parentState.ID(), updatedStateID)
+	require.Empty(s.T(), dbUpdates)
 }
 
-// TestMutatorHappyPathNoChanges tests that Mutator correctly indexes the protocol state when there are no changes.
-func (s *MutatorSuite) TestMutatorHappyPathNoChanges() {
+// TestHappyPathHasChanges tests that stateMutator returns cached db updates when building protocol state after applying service events.
+func (s *StateMutatorSuite) TestHappyPathHasChanges() {
+	s.params.On("EpochFallbackTriggered").Return(false, nil)
 	parentState := unittest.ProtocolStateFixture()
-	candidate := unittest.BlockHeaderFixture(unittest.HeaderWithView(parentState.CurrentEpochSetup.FirstView))
-	s.protocolStateDB.On("ByBlockID", candidate.ParentID).Return(parentState, nil)
-	updater, err := s.mutator.CreateUpdater(candidate)
-	require.NoError(s.T(), err)
+	s.stateMachine.On("ParentState").Return(parentState)
+	s.stateMachine.On("Build").Return(unittest.ProtocolStateFixture().ProtocolStateEntry,
+		unittest.IdentifierFixture(), true)
 
-	s.protocolStateDB.On("Index", candidate.ID(), parentState.ID()).Return(func(tx *transaction.Tx) error { return nil })
+	epochSetup := unittest.EpochSetupFixture()
+	result := unittest.ExecutionResultFixture(func(result *flow.ExecutionResult) {
+		result.ServiceEvents = []flow.ServiceEvent{epochSetup.ServiceEvent()}
+	})
 
-	err = s.mutator.CommitProtocolState(updater)(&transaction.Tx{})
+	block := unittest.BlockHeaderFixture()
+	seal := unittest.Seal.Fixture(unittest.Seal.WithBlockID(block.ID()))
+	s.headersDB.On("ByBlockID", seal.BlockID).Return(block, nil)
+	s.resultsDB.On("ByID", seal.ResultID).Return(result, nil)
+
+	s.stateMachine.On("ProcessEpochSetup", epochSetup).Return(nil).Once()
+	s.setupsDB.On("StoreTx", epochSetup).Return(func(*transaction.Tx) error { return nil }).Once()
+
+	err := s.mutator.ApplyServiceEvents([]*flow.Seal{seal})
 	require.NoError(s.T(), err)
+
+	_, _, _, dbUpdates := s.mutator.Build()
+	require.NoError(s.T(), err)
+	require.Len(s.T(), dbUpdates, 1)
```

Comment:
Should we add an assertion that `newStateID != parentStateID`?

Reply:
I didn't add it since `newStateID` is what we return from `Build` which is a fixture.

Reply:
Right, makes sense.